### ISSUE TITLE
history

### ISSUE THREAD

**Issue by rob-myers** [firehol/blocklist-ipsets#18]
Hi,

The commit history only goes up to September 25th, presumably because it is being restarted.
Is there a way to access a longer history, even for only some of the lists?

I ask because I'd like to cross reference with the scans at [censys.io](http://censys.io).

**Comment by ktsaou** [comment]
Hi, thanks for using github. It takes a tremendous amount of time to answer questions by email, so I just stopped doing it.

Regarding the history: unfortunately the change frequency is so high that after a week:

1. git structures occasionally are corrupted
2. every operation on the git repo is extremely slow
3. any attempt to compact the git structures takes more that a full day to be completed
4. it requires a lot of disk space to maintain it

So, I rebuild the git repo every week.

There are a few alternatives though depending on what you need. For example:

1. `update-ipsets`, the program that maintains the repo and the site http://iplists.firehol.org maintains always a snapshot of all incremental additions of the ipsets. This means that for each IP currently listed on any ipset, it knows the date and time it was added. Unfortunately it does not maintain a history of the removed IPs. So, when an IP has been removed from an ipset, its history is lost. I can provide the current history to you if you need it (~2GB on disk, ~200MB to download).

2. You can install and run `update-ipsets` yourself. If you do this, the program can maintain a git repo for you for all the history you need (given the issues I mentioned above) and it will also provide a site exactly like http://iplists.firehol.org on your web server.

**Comment by rob-myers** [comment]
I've installed update-ipsets, and intend to use it as you suggest -- thanks very much!

What do you mean by the "current history"?
The answer is probably 'yes' regardless.

**Comment by ktsaou** [comment]
nice!

current history = for an ip set that currently lists 100 IPs, to know when each of these IPs was added to the ipset - even if it is a year ago (I started in April 2015, so that is the oldest I know).

**Comment by ktsaou** [comment]
Since you installed `update-ipsets`, please enable the site. Once you do this, `update-ipsets` will use `/var/lib/update-ipsets` for maintaining its history. Test it a bit and see what it does in there.

Once you are comfortable with it, I can give you my `/var/lib/update-ipsets`, which has the **current history**. So, yours will continue from the point I have mine.

Keep in mind that `/var/lib/update-ipsets` is CPU specific. Mine is for intel/amd CPUs (both 32/64 bit is fine).

**Comment by rob-myers** [comment]
Hello again,
I still need to enable the site, which I shall try to do today.
May I ask why your dataset only tracks 100 addresses?

**Comment by ktsaou** [comment]
> May I ask why your dataset only tracks 100 addresses?

I am not sure I understand. What do you mean?

**Comment by rob-myers** [comment]
Well, my understanding is that "100 addresses" refers to 100 individual ip addresses, possibly subnets.
On the other hand, your site tracks a great many lists, many having upwards of 10,000 subnets.

Apologies if I'm misunderstanding things.

**Comment by ktsaou** [comment]
There is no such limit.
What lead you to conclude that there is such a limit? There might be a bug somewhere...

**Comment by rob-myers** [comment]
I am referring to your own statement in a previous comment:

>  current history = for an ip set that currently lists 100 IPs, to know when each of these IPs was added to the ipset - even if it is a year ago (I started in April 2015, so that is the oldest I know).

**Comment by ktsaou** [comment]
oh! ok. That was just an example. There is no limit.

**Comment by rob-myers** [comment]
Great, sorry for the confusion. I will press on...

**Comment by rob-myers** [comment]
Hi again,
I've gone through the wiki, except for the iptables stuff.
Would you mind providing the current history?

**Comment by ktsaou** [comment]
get it from: http://iplists.firehol.org/history-dumps/var_lib_update_ipsets_20161006.tar.bz2

unpack it somewhere and move it to `/var/lib/update-ipsets`.

This assumes you have already `update-ipsets` monitoring working (check the wiki).

Once you have it, let me know to delete it (limited space on that VM).

**Comment by rob-myers** [comment]
Got it, thanks very much!

**Comment by ktsaou** [comment]
ok. deleted it.

**Comment by rob-myers** [comment]
Hi, 

Given a monitored ipset, I can read the individual files e.g. via `iprange --print-single-ips`.

Does each file represent those IP addresses which were not in the previously downloaded version of the ipset?

**Comment by ktsaou** [comment]
Let's take an example: `dshield`.

```sh
# cd /var/lib/update-ipsets/dshield
# ls -l
total 2216
-rw-r--r-- 1 root root  449803 Oct  5 22:12 changesets.csv
-rw-r--r-- 1 root root    1625 Oct  5 22:12 histogram
-rw------- 1 root root     260 Jun 19  2015 histogram.old
-rw------- 1 root root  420476 Oct  5 22:12 history.csv
-rw------- 1 root root     253 Oct  5 22:11 latest
drwx------ 2 root root    4096 Oct  5 22:12 new
-rw------- 1 root root 1364366 Oct  5 22:12 retention.csv
```

`latest` is the latest version of the dshield ipset, in iprange binary format. To get the text version of it just run:

```sh
# iprange latest
46.161.40.0/24
46.172.91.0/24
61.240.144.0/24
71.6.146.0/24
71.6.216.0/24
74.82.47.0/24
91.224.160.0/23
94.102.49.0/24
123.31.34.0/24
141.212.122.0/24
163.172.67.0/24
183.60.48.0/24
185.35.62.0/24
185.56.82.0/24
185.93.185.0/24
198.20.69.0/24
204.42.253.0/24
209.126.122.0/24
222.186.21.0/24
```

The `new` directory, has a number of files:

```sh
# ls -l new/
total 48
-rw------- 1 root root 105 Sep  9 06:17 1473398250
-rw------- 1 root root 105 Sep  9 13:21 1473423697
-rw------- 1 root root 105 Sep 11 20:18 1473621532
-rw------- 1 root root 105 Sep 26 20:43 1474919032
-rw------- 1 root root 105 Oct  3 16:18 1475507931
-rw------- 1 root root 105 Oct  3 20:42 1475523730
-rw------- 1 root root 105 Oct  3 23:41 1475534495
-rw------- 1 root root 113 Oct  4 12:23 1475580228
-rw------- 1 root root 105 Oct  4 19:41 1475606493
-rw------- 1 root root 105 Oct  5 08:41 1475653308
-rw------- 1 root root 154 Oct  5 21:17 1475698625
-rw------- 1 root root 105 Oct  5 22:11 1475701888
```

For each of those, the filename is the unix timestamp it was added and the contents of it are the IPs/subnets that where added. So:

```sh
# date -d @1473398250
Fri Sep  9 06:17:30 BST 2016

# iprange new/1473398250
91.224.160.0/23
```

The sum of all the files in the directory `new` should be the same to `latest` files:

```sh
# iprange new/*
46.161.40.0/24
46.172.91.0/24
61.240.144.0/24
71.6.146.0/24
71.6.216.0/24
74.82.47.0/24
91.224.160.0/23
94.102.49.0/24
123.31.34.0/24
141.212.122.0/24
163.172.67.0/24
183.60.48.0/24
185.35.62.0/24
185.56.82.0/24
185.93.185.0/24
198.20.69.0/24
204.42.253.0/24
209.126.122.0/24
222.186.21.0/24
```

and to verify it:

```sh
# iprange new/* | iprange --header - --compare-next latest
name1,name2,entries1,entries2,ips1,ips2,combined_ips,common_ips
stdin,latest,19,19,5120,5120,5120,5120
```

They are the same, since `combined_ips` is the same with `common_ips`.

**Comment by ktsaou** [comment]
The verification can also be done with this:

```sh
# iprange new/* | iprange - --diff latest
<nothing printed>
```

**Comment by ktsaou** [comment]
The other files, are used for rendering the charts on the site.
What can be interesting for you is:

```sh
# head -n 10 changesets.csv
DateTime,IPsAdded,IPsRemoved
1434698783,5120,0
1434699761,3840,3840
1434701854,3328,3328
1434703338,3840,3840
1434705060,4096,4096
1434706942,3840,3840
1434708656,3840,3840
1434710543,4096,4096
1434711354,512,512
```

So, given a timestamp, it says how many IPs where added and removed.

Then:

```sh
# head -n 10 retention.csv
date_removed,date_added,hours,ips
1434699761,1434698783,0,3840
1434701854,1434699761,1,3328
1434703338,1434698783,1,256
1434703338,1434699761,1,256
1434703338,1434701854,0,3328
1434705060,1434699761,1,256
1434705060,1434703338,0,3840
1434706942,1434698783,2,256
1434706942,1434705060,1,3584
```

retention gives a number of IPs with the date added, removed, and time in hours it was listed.

```sh
# head -n 10 history.csv
DateTime,Entries,UniqueIPs
1433707032,20,5120
1433721409,20,5120
1433735821,20,5120
1433750202,20,5120
1433764555,20,5120
1433778957,20,5120
1433793352,20,5120
1433807755,20,5120
1433822078,20,5120
```

history is like a log (timestamp, entries in ipset, IPs in ipset)

Finally, histogram has 2 BASH associative arrays for the site charts.

**Comment by rob-myers** [comment]
Very useful examples, thanks.

Each individual file in `new/` represents a collection of IP addresses and has a creation date.
How is this collection computed from the ipset?

Also, since addresses may be removed from ipsets, surely `latest` is not always the union of `new/*`?

**Comment by ktsaou** [comment]
> How is this collection computed from the ipset?

When a new version is downloaded from the maintainer, the latest and the new version downloaded are compared. The new IPs added are saved in `new/`. The IPs removed are not tracked (they are lost). Once this completes, the newly downloaded version become the `latest`.

So, `new/*` and `latest` should always be the same (there is just a fraction of time, where this is not true, while `update-ipsets` is running and updating the given ipset).

**Comment by rob-myers** [comment]
Understood.
So, even though the removal date is lost, the event `added > removed > added` is detectable.

**Comment by ktsaou** [comment]
Let me rephrase it:

When a new version is downloaded from the maintainer:

1. the latest and the new version downloaded are compared.
2. The new IPs added are saved in `new/` with the timestamp of the latest file downloaded.
3. For each file already in `new/`, a check is made if it lists IPs that have been removed. It it lists any removed IPs, it is re-written to list only the IPs currently listed.
4. The IPs removed are not tracked (they are lost).
5. Once this completes, the newly downloaded version become the `latest`.

I added step 3.

**Comment by rob-myers** [comment]
Thanks, I also had a look through `update-ipsets.in` inside `retention_detect`.

Did you choose to delete the removed IPs to save space?
I'm intending to run your great tool regularly, but I'll be changing that behaviour.

**Comment by ktsaou** [comment]
> Did you choose to delete the removed IPs to save space?

yes, exactly.

This is the line that deletes the removed IPs file: https://github.com/firehol/firehol/blob/4597f02c629b2bed7bfd0a7efec41ec8201a5b4f/sbin/update-ipsets.in#L1682

At that point, the file is stored in a tmp dir. To keep it, you have to move it somewhere.
